packages/back-end/src/services/datasource.ts

```diff
@@ -82,6 +82,42 @@ export function getSourceIntegrationObject(datasource: DataSourceInterface) {
   return obj;
 }
 
+export async function generateInformationSchema(
+  datasource: DataSourceInterface
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): Promise<any> {
```

Comment:
I need to refactor this as currently the return types are all over the place, and it's making it difficult to make Typescript happy, which is usually a sign something isn't very stable.

Currently, only BigQuery and Postgres will have the `getInformationSchema` method, so I need a way to exit out of this method gracefully if it's called from a datasource other than the two above. This way, if someone goes to add a datasource other than BigQuery or Postgres, it won't break. 

However, I don't want to write a manual check like `if (datasource.type !== "BigQuery" || "Postgres") return, as that will get very verbose down the road when say, all but 1 of the datasource types support building a schema. So, I thought I'd add a check where if the integration doesn't contain the method `runGetSchemaQuery` then we'd return gracefully.

This works, but it makes the return type really complicated, especially since below, we're also returning either `informationSchema` or `error`.

All in all, it feels like I need to change this pattern.

One thing I tried was lifting the `getSourceIntegrationObject` out of this, so I could check if the integration had the `getInformationSchema` method and only call `generateSchema` if it existed. But it feels weird to repeat myself so many times.

I also attempted to define the crazy return (that's commented out) but that still doesn't make Typescript happy.

Reply:
What if you added `getInformationSchema()` to the base class `SqlIntegration` and had that return type be `null | InformationSchema`? I'm not sure what an information schema looks like yet but it looks like an array of results.

Then, whenever you go to call `integration.getInformationSchema()` you can check if it's null or not, and if it's null don't use it.

You could have `InformationSchema` implement a union type of `PostgresInformationSchema | BigQueryInformationSchema` assuming you know the shapes of each. 

So your method could be:

```ts
async getInformationSchema(): Promise<null | InformationSchema> {
  // ...
}
```

And the types:

```ts
type PostgresInformationSchema = {
  type: 'postgres';
  schema: {
    // todo: define the schema
  }
}

type BigQueryInformationSchema = {
  type: 'bigquery';
  schema: {
    // todo: define the schema
  }
}

type InformationSchema = |
  BigQueryInformationSchema | 
  PostgresInformationSchema;
```

Then if you needed to look at the schema for whatever reason, you could do a check:

```ts
if (!informationSchema) {
  return;
}

switch (informationSchema.type) {
  case 'bigquery':
    return handleBigQueryInformationSchema(informationSchema);
  case 'postgres':
    return handlePostgresInformationSchema(informationSchema);
}
```

Those above-mentioned methods can safely use the specific information schema type, e.g.:

```ts
export const handlePostgresInformationSchema = (informationSchema: PostgresInformationSchema) => {
  // ...
}
```

And then don't add a `default` statement to your switch so that when you add a new `InformationSchema`, you can ensure you cover all cases that need one.

That's generally how I've handled variable things like this, for example, the [`NotificationEvent`](https://github.com/growthbook/growthbook/blob/7e11e7de6a1b8d4884cdbc8a8038fbbc1b18aadf/packages/back-end/src/events/base-events.ts#L66) which populates the events collection.

Reply:
@tinahollygb Thank you for your suggestion! I was going through this, and it sparked one idea that is a bit of a smaller change. 

In the event this is for a datasource type that doesn't support the `informationSchema` rather than the method returning `undefined` I'm just returning `{ informationSchema: [] }`. And then I'm checking the length of `informationSchema` - I suppose that isn't required, but this way, the informationSchema key will only exist if there is an actual informationSchema, which should make checks on the front-end a bit easier.

Let me know what you think of this approach.